pjmedia/src/pjmedia-codec/and_vid_mediacodec.cpp

```diff
@@ -0,0 +1,1908 @@
+    const int code_size = PJ_ARRAY_SIZE(start_code);
+    unsigned buf_pos, whole_len = 0;
+    unsigned i, frm_cnt;
+    pj_status_t status;
+
+    PJ_ASSERT_RETURN(codec && count && packets && out_size && output,
+                     PJ_EINVAL);
+    PJ_ASSERT_RETURN(output->buf, PJ_EINVAL);
+
+    and_media_data = (and_media_codec_data*) codec->codec_data;
+
+    /*
+     * Step 1: unpacketize the packets/frames
+     */
+    whole_len = 0;
+    if (and_media_data->whole) {
+	for (i=0; i<count; ++i) {
+	    if (whole_len + packets[i].size > and_media_data->dec_buf_size) {
+		PJ_LOG(4,(THIS_FILE, "Decoding buffer overflow [1]"));
+		return PJMEDIA_CODEC_EFRMTOOSHORT;
+	    }
+
+	    pj_memcpy( and_media_data->dec_buf + whole_len,
+	               (pj_uint8_t*)packets[i].buf,
+	               packets[i].size);
+	    whole_len += packets[i].size;
+	}
+
+    } else {
+	h264_codec_data *h264_data = (h264_codec_data *)and_media_data->ex_data;
+
+	for (i=0; i<count; ++i) {
+
+	    if (whole_len + packets[i].size + code_size >
+                and_media_data->dec_buf_size)
+	    {
+		PJ_LOG(4,(THIS_FILE, "Decoding buffer overflow [2]"));
+		return PJMEDIA_CODEC_EFRMTOOSHORT;
+	    }
+
+	    status = pjmedia_h264_unpacketize( h264_data->pktz,
+					       (pj_uint8_t*)packets[i].buf,
+					       packets[i].size,
+					       and_media_data->dec_buf,
+					       and_media_data->dec_buf_size,
+					       &whole_len);
+	    if (status != PJ_SUCCESS) {
+		PJ_PERROR(4,(THIS_FILE, status, "Unpacketize error"));
+		continue;
+	    }
+	}
+    }
+
+    if (whole_len + code_size > and_media_data->dec_buf_size ||
+    	whole_len <= code_size + 1)
+    {
+	PJ_LOG(4,(THIS_FILE, "Decoding buffer overflow or unpacketize error "
+			     "size: %d, buffer: %d", whole_len,
+			     and_media_data->dec_buf_size));
+	return PJMEDIA_CODEC_EFRMTOOSHORT;
+    }
+
+    /* Dummy NAL sentinel */
+    pj_memcpy(and_media_data->dec_buf + whole_len, start_code, code_size);
+
+    /*
+     * Step 2: parse the individual NAL and give to decoder
+     */
+    buf_pos = 0;
+    for ( frm_cnt=0; ; ++frm_cnt) {
+	pj_uint32_t frm_size;
+	pj_bool_t write_output = PJ_FALSE;
+	unsigned char *start;
+
+	for (i = code_size - 1; buf_pos + i < whole_len; i++) {
+	    if (and_media_data->dec_buf[buf_pos + i] == 0 &&
+		and_media_data->dec_buf[buf_pos + i + 1] == 0 &&
+		and_media_data->dec_buf[buf_pos + i + 2] == 0 &&
+		and_media_data->dec_buf[buf_pos + i + 3] == 1)
+	    {
+		break;
+	    }
+	}
+
+	frm_size = i;
+	start = and_media_data->dec_buf + buf_pos;
+	write_output = (buf_pos + frm_size >= whole_len);
+
+	status = and_media_decode(codec, and_media_data, start, frm_size, 0,
+		              &packets[0].timestamp, write_output, output);
+	if (status != PJ_SUCCESS)
+	    return status;
+
+	if (write_output)
+	    break;
+
+	buf_pos += frm_size;
+    }
+
+    return PJ_SUCCESS;
+}
+
+static pj_status_t vpx_unpacketize(struct and_media_codec_data *and_media_data,
```

Comment:
if there are a lot of similarities with libvpx, perhaps better to refactor it to another file, similar to h263/h264_packetizer?